add --exclude option

aymanosman opened the pull request:
I often need to use the `--exclude` option. This PR adds support for it.

Thanks to transient, it also supports supplying multiple `--exclude` options and the values supplied can even be saved for later invocations by `transient-set` usually bound to `C-x s` (See [Saving Values](https://www.gnu.org/software/emacs/manual/html_mono/transient.html#Saving-Values)).

<img width="800" alt="Screenshot 2023-11-17 at 13 06 11" src="https://github.com/ananthakumaran/exunit.el/assets/3605049/e94f00b5-7fcd-4c03-b785-c64661143710">

ananthakumaran commented:
Sorry about the late reply. I haven't looked into the details, is there any way we can make a transient definition user configurable?

aymanosman commented:
> Sorry about the late reply. I haven't looked into the details, is there any way we can make a transient definition user configurable?

I'm not too familiar with how transient works and I'm not sure what kind of configurability you are suggesting. If you tell me what kind of user customisation you'd like to see, I could investigate whether it is possible.

ananthakumaran commented:
> I'm not sure what kind of configurability you are suggesting.

Let's take the current example, you are proposing to add `--exclude`, we have only fixed space available on UI, so the package has to take a call on whether to include the flag or not. If that could be done by the user directly, that would solve the UI space problem. Basically, as a User, I should be able to add my own flags.

aymanosman commented:
Cool, I’ll see if this can be done with transient.

aymanosman commented:
@ananthakumaran So, there are a couple of options of how to implement this which I got from [this discussion on reddit](https://www.reddit.com/r/emacs/comments/1agq74e/how_to_make_a_transient_prefix_optional_user/?utm_source=share&utm_medium=web2x&context=3)

1. Use the `:if` parameter of a transient command

A patch would look like this:

```diff
diff --git a/exunit.el b/exunit.el
index 1b61c4d..009b66b 100644
--- a/exunit.el
+++ b/exunit.el
@@ -39,12 +39,18 @@

 ;;; Private

+(defcustom exunit-transient-extra-flags '(--exclude)
+  "Controls the extra flags available in the transient interface"
+  :type '(repeat string)
+  :group 'exunit)
+
 (transient-define-infix exunit-transient:--exclude ()
   :description "Exclude"
   :class 'transient-option
   :multi-value 'repeat
   :shortarg "-e"
-  :argument "--exclude=")
+  :argument "--exclude="
+  :if (lambda () (memq '--exclude exunit-transient-extra-flags)))

 (transient-define-prefix exunit-transient ()
   "ExUnit"
```

2. Set the "level" of the command above the `transient-default-level`

The patch would look something like this:

```diff
diff --git a/exunit.el b/exunit.el
index 1b61c4d..bb48d6b 100644
--- a/exunit.el
+++ b/exunit.el
@@ -52,10 +52,11 @@
    [("-f" "Failed" "--failed")
     ("-s" "Stale" "--stale")
     ("-t" "Trace" "--trace")
-    ("-c" "Coverage" "--cover")
-    (exunit-transient:--exclude)]
+    ("-c" "Coverage" "--cover")]
    [("-z" "Slowest" "--slowest=10")
-    ("-m" "Fail Fast" "--max-failures=1")]]
+    ("-m" "Fail Fast" "--max-failures=1")]
+   [5
+    (exunit-transient:--exclude)]]
   ["Actions"
    [("a" "all" exunit-verify-all)
     ("v" "current buffer" exunit-verify)]
``` 

I'm not that familiar with this part of transient, and I'm not sure how familiar most users are with it either.

ananthakumaran commented:
Thanks for checking, I am leading towards using [level](https://www.gnu.org/software/emacs/manual/html_mono/transient.html#Enabling-and-Disabling-Suffixes) since that seems to be designed to solve this exact issue. I understand there is a discoverability issue, but we could put a note about the transient level. Let me know what you think. If you agree go ahead and update the PR, I will test it.

aymanosman commented:
I went ahead and implemented the change using the "levels" functionality. I added a note in the readme that shows users how to make the extra `--exclude` switch visible.
